### PR TITLE
fix issues #187 #188 and #189 and better testing for StatementSorter

### LINKED ISSUE #187: SHACL generation sometime produces duplicate triples
Since triples with blank nodes are canonicalized they should be cleaned if doubles appear.

### LINKED ISSUE #188: Calculation of canonical blank nodes has some issues in StatementSorter
Treatment of blank nodes and their unique hashcode has some issues. Calculation for blank nodes in `subject` location must be different in case they are root (i.e. not referenced in an `object`) or not root (i.e. referenced by an `object`).

### LINKED ISSUE #189: Make lists in SHACL generation more predictable
List of languages for `sh:languageIn` and list of datatypes can be sorted so they will always have the same order.

### LINKED ISSUE #191: SHACL generator can now work with reuse of blank nodes check
Should be removed

### PR DESCRIPTION
**Pull request by natan-cox-cognizone**

### CODE REVIEW

File: cube/src/main/java/zone/cogni/asquare/cube/digest/SortedBlock.java [hunk 1]
```diff
@@ -139,4 +157,18 @@ public List<SortedBlock> getNestedBlocks() {
     return nestedBlocks == null ? Collections.emptyList()
                                 : Collections.unmodifiableList(nestedBlocks);
   }
+
+  public String toString() {
+    return toString(0);
+  }
+
```
Comment (natan-cox-cognizone) [inline cube/src/main/java/zone/cogni/asquare/cube/digest/SortedBlock.java:137]: `toString` for better debugging.

File: cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java [hunk 2]
```diff
@@ -34,24 +36,38 @@
...
+      boolean isSame = getStatementComparator().compare(statement, nextStatement) == 0;
+      if (isSame) {
+        statements.remove(i + 1);
+        i -= 1;
+      }
+    }
+
+    return statements;
+  }
+
```
Comment (natan-cox-cognizone) [inline cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java:70]: Detection and removal of duplicates.

File: cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java [hunk 3]
```diff
@@ -63,20 +79,53 @@ private void processBlock(List<Statement> statements, SortedBlock parent, Sorted
...
-    RDFNode object = statement.getObject().isAnon() ? getDigestBlankNode(currentBlock)
-                                                    : statement.getObject();
-    return ResourceFactory.createStatement(subject,
-                                           statement.getPredicate(),
-                                           object);
+    Resource subject = getConvertedSubject(parentBlock, statement);
+    RDFNode object = getConvertedObject(currentBlock, statement);
+    return ResourceFactory.createStatement(subject, statement.getPredicate(), object);
+  }
+
```
Comment (natan-cox-cognizone) [inline cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java:86]: logic for correct blank node id calculation

File: cube/src/main/resources/convertor/data2shacl/select-datatype.sparql.spel [hunk 4]
```diff
@@ -6,4 +6,5 @@ select distinct ?value {
 
   bind (datatype(?x) as ?datatypeValue)
   bind (coalesce(?datatypeValue, rdfs:langString) as ?value)
-}
\ No newline at end of file
+}
+order by ?value
```
Comment (natan-cox-cognizone) [inline cube/src/main/resources/convertor/data2shacl/select-datatype.sparql.spel:10]: fix for sorting languages.

File: cube/src/test/java/zone/cogni/asquare/cube/sort/StatementSorterTest.java [hunk 5]
```diff
@@ -3,69 +3,165 @@
...
 
 import java.util.List;
+import java.util.stream.Collectors;
 
+import static org.apache.jena.rdf.model.ResourceFactory.createStringLiteral;
 import static org.assertj.core.api.Assertions.assertThat;
 
 class StatementSorterTest {
 
+
```
Comment (natan-cox-cognizone) [inline cube/src/test/java/zone/cogni/asquare/cube/sort/StatementSorterTest.java:24]: lots of new and much simpler tests... those mixed tests were actually already to complex to correctly evaluate

File: cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java (hunk 2 above)
Comment (michielrogissart-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java:70]: Wouldn't casting to set work as well? 🤔

Comment (natan-cox-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java:70]: well. then we loose order?

Comment (michielrogissart-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java:70]: I'm wondering: does it make sense to modify the argument and return it? Shouldn't you either return a new list or modify the argument in a `void` function? Is very nitpicky, I admit :)

**Review by michielrogissart-cognizone — APPROVED** [review]

File: cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java (hunk 2 above)
Comment (natan-cox-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/sort/StatementSorter.java:70]: true. I was not sure either but decided to return it so we can change the implementation later. @michielrogissart-cognizone

**Pull request merged**